### DISCUSSION TITLE
Describe/explain why different Jinja features are/aren't available in different contexts

### DISCUSSION THREAD

**Discussion by joellabes** [dbt-labs/docs.getdbt.com#1049] [Archive]
This should answer questions like: 
- "Why can't I use a var in a var in `dbt_project.yml`?"
- "Why can I do `materialized: "{{ 'table' if target.name == 'prod' else 'view' }}"` in `dbt_project.yml` but not 
`materialized: {{ table_if_prod() }}`?"
- Why does `{{ this }}` sometimes evaluate to None?

With answers like 
- vars are defined _inside_ `dbt_project.yml`, so you can't access them until after the whole file has been evaluated, which is too late.
- Because `target` is evaluated before `dbt_project.yml` early and macros aren't available until _after_
- I don't know offhand if this is true, but I bet it is and good lists come in threes. Maybe ephemeral models?

It would be useful to show the order of parsing, describe the concepts, and show a matrix of where things can be used.

I think this is a matrix of 

*Different tools*
- `var`
- `this` 
- `target`
- custom macros 
- ...

vs *different contexts*

- `dbt_project.yml` 
- `config()` block
- inside a model's body
- ...

**Comment by runleonarun** [comment]
@joellabes Is there someone or a group of someones we should @mention to get this discussion rolling?

**Reply by joellabes** [reply]
Unfortunately for his inbox, I think @jtcohen6 is probably still the person with the canonical list in his head 😬

**Reply by jtcohen6** [reply]
The _canonical_ canonical list is: https://github.com/dbt-labs/dbt-core/tree/main/core/dbt/context

There's an inline comment [here](https://github.com/dbt-labs/dbt-core/blob/13b18654f03d92eab3f5a9113e526a2a844f145d/core/dbt/context/base.py#L28-L59) that's helpful, but really we need a README for this submodule! Then we could link to it from this docs page? (cc @dbt-labs/core-language)
